xdsl/backend/riscv/register_rvscf_regalloc.py

```diff
@@ -0,0 +1,201 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+from xdsl.dialects import riscv, riscv_func, riscv_scf
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext, Region, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+@dataclass
+class RegAllocCtx:
+    """
+    This is the context object for allocating registers.
+
+    It keeps track of how many times each ssa value is expected to be seen before
+    it is "free" again.
```

Comment:
@AntonLydike and @adutilleul, do we indeed need a context object. AFAIU, if we allocate backwards we do not need to maintain use counts, no?